// eslint-plugin-solid/dist/rules/jsx-uses-vars.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@typescript-eslint/utils");
const createRule = utils_1.ESLintUtils.RuleCreator.withoutDocs;
exports.default = createRule({
    meta: {
        type: "problem",
        docs: {
            recommended: "error",
            description: "Prevent variables used in JSX from being marked as unused.",
            url: "https://github.com/solidjs-community/eslint-plugin-solid/blob/main/docs/jsx-uses-vars.md",
        },
        schema: [],
        messages: {},
    },
    defaultOptions: [],
    create(context) {
        return {
            JSXOpeningElement(node) {
                let parent;
                switch (node.name.type) {
                    case "JSXNamespacedName":
                        return;
                    case "JSXIdentifier":
                        context.markVariableAsUsed(node.name.name);
                        break;
                    case "JSXMemberExpression":
                        parent = node.name.object;
                        while ((parent === null || parent === void 0 ? void 0 : parent.type) === "JSXMemberExpression") {
                            parent = parent.object;
                        }
                        if (parent.type === "JSXIdentifier") {
                            context.markVariableAsUsed(parent.name);
                        }
                        break;
                }
            },
            "JSXAttribute > JSXNamespacedName": (node) => {
                var _a, _b;
                if (((_a = node.namespace) === null || _a === void 0 ? void 0 : _a.type) === "JSXIdentifier" &&
                    node.namespace.name === "use" &&
                    ((_b = node.name) === null || _b === void 0 ? void 0 : _b.type) === "JSXIdentifier") {
                    context.markVariableAsUsed(node.name.name);
                }
            },
        };
    },
});
